### PR TITLE
add blog

### PR DESCRIPTION
**Pull request by jwdashel**

### CODE REVIEW

File: docs/blog/unettest.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
+# UNETTEST
+
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:2]: We probably want a title that's more descriptive of the problem being solved. Something like "Using TDD to Learn Unfamiliar Frameworks" or something to that effect.

File: docs/blog/unettest.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+![What is the answer, O Deep Thought](https://docs-unettest.s3.us-east-2.amazonaws.com/o_deep_thought.png)
+
+When I write software, there are many ways to ask a question. I could go to
+Stack Overflow or a smarter coworker, but I usually like to ask my computer
+first.  What do I do if I feel like my computer is making me wait [seven and a
+half million years](https://hitchhikers.fandom.com/wiki/Deep_Thought) to get an
+answer? 
+
+I would want to revisit the question of how to ask questions. If I want to know
+the Ultimate Answer to Life, the Universe, and Everything, I must ask a damn
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:12]: replace "a damn" with "the". Let's keep the language professional even if we don't speak like that in person.

File: docs/blog/unettest.md [hunk 3]
```diff
@@ -0,0 +1,288 @@
+# UNETTEST
+
+![What is the answer, O Deep Thought](https://docs-unettest.s3.us-east-2.amazonaws.com/o_deep_thought.png)
+
+When I write software, there are many ways to ask a question. I could go to
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:5]: When writing blog posts for NYPR, we want to use "we" instead of "I". This is going to require some doing, but it's important that we keep a consistent voice across all of our blog posts.

File: docs/blog/unettest.md [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+of effort to calculate.
+
+42 is an interesting data point and it might be useful, but in order to figure
+out what's going on, I might also like to know the answer to Death, the
+Universe, and Everything. And I might like to ask a question about just the
+Universe on its own, maybe a question or two on that Everything bit, and perhaps
+a couple of questions about a bowl of petunias. I cannot afford to wait so long
+between each answer. I have things to do.
+
+I do not have time to build the Earth in order to ask my questions so instead I
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:23]: I'm going to be honest here and say I don't think you need the Hitchhiker's analogy. Even though it's fun, it distracts from the main point of the post. I think what is more important is making sure that your audience knows what TDD is and what NGINX is.

File: docs/blog/unettest.md [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+NonProd and wave around my dowsing rod to try and sense if things are both
+working and not broken. High cost, low reliability.
+
+This is silly in the best of times, but workable--no reason to fix what ain't
+broke. Until I started scratching around the corners of an NGINX edge case...
+The only way I was able to figure out what on Earth was going on was by walking
+through all the different potential paths of logic, and ugh, that was not
+pretty. Let's get into it.
+
+## The Monolith and the Microservice
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:48]: This section definitely has the right tone, and gets to the point, this is what I was looking for up above.

File: docs/blog/unettest.md (hunk 3 above)
Comment (ncastro-nypr) [reply docs/blog/unettest.md:5]: in the next section, ##The Monolith and the Microservice, you use "we" and "I" much better. I don't mind it there as much so try to incorporate that voice up here in the intro.

File: docs/blog/unettest.md [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+I was writing NGINX configurations some time ago. I was new to NGINX and trying
+to figure it out. Their documentation (lovely though it is) can be impenetrable.
+I was grokking the config language by looking at existing configurations and
+mimicking what they do. I could not just hack at it though because this bouquet
+of CONF files is handling all of our network traffic--it is largely a
+look-don't-touch affair. If I had a question, I could not ask it of my computer
+easily. NGINX does not have support for unit tests (to my knowledge), or rapid
+feedback of any kind. If I want to know the answer to a question, I cannot
+write a test and run it locally. I would have to deploy that question to
+NonProd and wave around my dowsing rod to try and sense if things are both
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:39]: use the term "Staging" instead of "NonProd", it'll be less confusing.

File: docs/blog/unettest.md (hunk 3 above)
Comment (ncastro-nypr) [reply docs/blog/unettest.md:5]: If you need more explicit help on how to convert everything, lmk

File: docs/blog/unettest.md [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+configuration changes. Sometimes the test results would not appear at all,
+sometimes I would be searching old time frames, sometimes old tests would be
+mixed in with the current tests, I was flooded with unrelated logs. It was
+bad.
+
+It was a bad way of asking questions. This development cycle was so kludgy , I
+was able to suss out WHAT the behavior was but not WHY. Not even WHAT EXACTLY.
+
+## Life, the Universe, and Everything
+
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:128]: In this section, you revert back to the "I" voice. I think the change comes whenever you go from explaining a technical issue to telling a story. If you think about successful tech blog posts you've read, which of these two components is more compelling for you? Most likely the explanation of the technical issue. Personally, I enjoy reading your story, but for a wider audience, it's going to dilute what they're here for.

**Comment by ncastro-nypr** [comment]
So in general, I'm asking for some pretty major changes here. We can chat over slack or zoom or w/e to have a more nuanced discussion about my points if you'd like.

File: docs/blog/unettest.md [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+This yaml config tells `unettest` what the network looks like: what services are
+on the network and what their interfaces look like. Now we can develop
+confidently against them.
+
+Things we do not have to worry about:
+
++ that Monolith does 10 million other things
++ what Monolith or Microservice does when these particular endpoints are invoked
++ that Microservice is a Lambda or that Monolith is a legacy pile of spagetti
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:198]: spaghetti

File: docs/blog/unettest.md [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+![What is the answer, O Deep Thought](https://docs-unettest.s3.us-east-2.amazonaws.com/o_deep_thought.png)
+
+When we write software, there are many ways to ask a question. We could go to
+Stack Overflow or to a smarter coworker, but I like to ask my computer first.
+There are many ways to ask computers questions, and one of the most effective is
+with TDD. TDD is awesome. We can ask many questions quickly, and once a question
+is asked, we can be confident it will remain answered. Given a question on how
+something should work, write a test to describe the behavior. Then run the tests
+and write code to make the test green. Great stuff. If there is a corner case, a
+test case can expose it. When code turns red, make it better.
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:12]: This paragraph feels a little to me like the intro paragraph of an essay. Analogous to "At the dawn of man, good an evil was born...". Too generic and devoid of substance. 
My advice would be to scrap the whole thing and add in the meat of this paragraph below your next paragraph. I'll specify in another comment.

File: docs/blog/unettest.md [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+
+I was writing NGINX configurations some time ago. I was new to NGINX and trying
+to figure it out. Their documentation (lovely though it is) can be impenetrable.
+I was grokking the config language by looking at existing configurations and
+mimicking what they do. If I had a question, I could not ask it of my computer
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:17]: because I'm asking you to scrap the first paragraph, you'll have to rework your sentences about having questions since the context is now gone.

File: docs/blog/unettest.md [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+easily. NGINX does not have support for unit tests (to my knowledge), or rapid
+feedback of any kind. If I want to know the answer to a question, I cannot write
+a test and run it locally. I would have to guess and deploy to Staging then wave
+around my DevOps dowsing rod to try and sense if things are both working and not
+broken. High cost, low reliability.
+
+This is silly but it was working more or less--until I started scratching around
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:24]: "it was working" is a bit ambiguous
```suggestion
Despite the terrible workflow, I made due--until I started scratching around
```

File: docs/blog/unettest.md [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+the corners of an NGINX edge case... The only way to figure out what on Earth
+was going on was to walk through all the different potential paths of logic, and
+ugh, that was not pretty. Let's get into it.
```
Comment (ncastro-nypr) [inline docs/blog/unettest.md:27]: Per my comment above, I think you can add some words about TDD here. It would seem to flow a little better IMO.
```suggestion
ugh, that was not pretty. Rather than having to rely on the slow and fragile process of testing my changes in a staging environment after incurring the cost of deploying them, testing my changes locally would be far safer and faster. But without a pre-existing testing framework provided by NGINX, I'd have to write my own. The design for it would need to enable me to practice TDD (which you can read about here: [link to some article describing TDD to your satisfaction]). Let's get into it.
```

Comment (ncastro-nypr) [reply docs/blog/unettest.md:27]: this change also does something important - it says what the point of this blog post is. You wrote your own testing framework for NGINX, but don't mention it until way later in the post.

File: docs/blog/unettest.md [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+```
+location ~* /microservice_mirror/(.*) {
+         internal;
+         proxy_pass $microservice/new/api/update?value=$1;
+
+         ...some other configurations
+}
+```
+
+> Note: I am trying to do a little funny business with the `$1`. The format of
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md:80]: I think the tense _should_ change in the context of the note block. Since I am describing what I *am* trying to do in the code (not the narrative)

File: docs/blog/unettest.md [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+Docker, how would it talk to my services? I was sick of trying to hook into
+real-world networks and real-world services. There were too many complications,
+side effects, and fragile bits. I wanted to rigorously isolate my NGINX files.
+They wanted to squirm away. I was bent on pinning them down for interrogation.
+
+![Interrogation](https://docs-unettest.s3.us-east-2.amazonaws.com/rick+dicker+interrogation.png)
+
+At this point we are only interested in the Monolith and the Microservice
+abstractly, as **interfaces**. We do not care what these services do after they
+have been sent the correct request (that's a side effect). Only that they _have_
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md:171]: the side effect is a crucial piece of information here and i want to call it out in the parenthetical aside

File: docs/blog/unettest.md [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+Why not mock them?
+
+The only thing matters about Monolith here is that it is called with
+`/api/v1/important_update/mycrazyval` and the only thing important about the
+Microservice is that it is invoked at the same time but with
+`/new/api/update?value=mycrazyval`.
+
+If we were writing code, we would mock these external systems. Why should
+network rules be exempt? Let's Mock those services!
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md:182]: References to code should be in the present tense. I want to mark that transition.

File: docs/blog/unettest.md [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
++ what Monolith or Microservice does when these particular endpoints are invoked
++ that Microservice is a Lambda or that Monolith is a legacy pile of spagetti
++ what Monolith's dependencies and requirements are beyond the scope of its public
+interface
++ that these might be black-box services that cannot be controlled or might (!!)
+be only available in Production environments 
+
+Under the hood, `unettest` will parse this valid and complete configuration file
+and create two little local containerized web services conforming to these
+definitions. They are not functional, but they behave as if they are and their
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md:220]: as above, i want all code (something someone might reasonably follow along with on their own computer) to be present tense

File: docs/blog/unettest.md [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+
++ that Monolith does 10 million other things
++ what Monolith or Microservice does when these particular endpoints are invoked
++ that Microservice is a Lambda or that Monolith is a legacy pile of spagetti
++ what Monolith's dependencies and requirements are beyond the scope of its public
+interface
++ that these might be black-box services that cannot be controlled or might (!!)
+be only available in Production environments 
+
+Under the hood, `unettest` will parse this valid and complete configuration file
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md:218]: noted tense change but changed to present not past

File: docs/blog/unettest.md [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+easily write a lambda plugin (or ask me! I take feature requests!) to run your
+lambda on your laptop. Amazon has announced SAM Local, but what I've read, it
+looks too heavy for me. It didn't work well with my current CI setup.  I don't
+want to replace my deployment infrastructure to run a lambda on localhost. But
+also it is annoying to write around the specialist AWS infrastructure (hello,
+lambda specific request json--`statusCode`, `body`, `content`) just to run a
+service. `unettest` can jack up your Lambda or NGINX configuration or anything
+else into a test harness so you can easily get under the hood of your network.
+See how things really work in the isolated safety of your local dev setup. u got
+a network? u net test.
```
[the remark this answers is not in the document]
Comment (jwdashel) [reply docs/blog/unettest.md]: i feel great about them :)

File: docs/blog/unettest.md [hunk 19]
```diff
@@ -0,0 +1,317 @@
+# Network Rules as (Tested) Code
+
+![What is the answer, O Deep Thought](https://s3.amazonaws.com/unettest.net/assets/o_deep_thought.png)
+
+Sometimes as developers, we start a ticket thinking it will be easy. Sometimes
+it turns out to be not easy. The natural thing to do is to reach for our trusted
+tools, those reliable methods and strategies for digging ourselves out of holes
+like this. And sometimes, those tools are missing.
+
+So we roll our own.
```
Comment (dbuvanova) [inline docs/blog/unettest.md:10]: roll out

**Comment by jjnypr** [comment]
👍
